### PR TITLE
Upstream changes

### PR DESCRIPTION
**Pull request by flatcap**
Here's a set of self-contained upstream changes that are ready to be merged.

### CODE REVIEW

File: bcache.c [hunk 1]
```diff
@@ -49,26 +49,23 @@ char *C_MessageCachedir; ///< Config: (imap/pop) Directory for the message cache
...
  */
-static int bcache_path(struct ConnAccount *account, const char *mailbox, char *dst, size_t dstlen)
+static int bcache_path(struct ConnAccount *account, const char *mailbox, struct BodyCache *bcache)
 {
   char host[256];
   struct Url url = { U_UNKNOWN };
-  int len;
 
-  if (!account || !C_MessageCachedir || !dst || (dstlen == 0))
+  if (!account || !C_MessageCachedir || bcache)
```
Comment (Austin-Ray) [inline bcache.c]: The conditional is going the wrong way. Should be `!bcache` https://github.com/neomutt/upstream-mutt/commit/ad72ec6c#diff-dd9883995d9854abd43600c8140f2d3bR46

Comment (flatcap) [reply bcache.c]: Argh!  Thanks.

**Comment by Austin-Ray** [comment]
commit 59c78c3fd6160a06effc492bd4f47aed1402cf33 looks good. only 7 more to go :)

**Comment by Austin-Ray** [comment]
Commit a6aadc433c09461c96dcf3d587f205b107f6d3e0 looks good. 6 more to go.

**Comment by Austin-Ray** [comment]
Commit b2eedddb737111e3c2308e573eda1de57de8be34 looks good. Half way there :)

File: curs_lib.c [hunk 2]
```diff
@@ -243,8 +242,8 @@ struct KeyEvent mutt_getch(void)
  * @retval 0  Selection made
  * @retval -1 Aborted
  */
-int mutt_get_field_full(const char *field, char *buf, size_t buflen, CompletionFlags complete,
-                        bool multiple, char ***files, int *numfiles)
+int mutt_buffer_get_field_full(const char *field, struct Buffer *buf, CompletionFlags complete,
+                               bool multiple, char ***files, int *numfiles)
```
Comment (Austin-Ray) [inline curs_lib.c]: Yikes that `files` variable is a nasty type!

**Comment by Austin-Ray** [comment]
Commit df97a8585a2e3ef6d45ff0d10f29731b39ca249a looks good

**Comment by Austin-Ray** [comment]
Commit 3677c4b0bde1ba49a0b05281bfae191f067bd590 looks good

**Comment by Austin-Ray** [comment]
Commit 6a3420396ef3a6c038873e693eb4502d0e1a1769 looks good

**Comment by flatcap** [comment]
> I didn't know if we wanted to use the `mutt_buffer_is_empty()` helper function

Ah, good spot.  That actually snuck in 6 months ago.
I've added an extra commit to fix that and 8 others.

Thanks.

**Pull request merged**